Add Contribution Guidelines

bnb opened the pull request:
Added the Contribution Guidelines to CONTRIBUTING.md, with sections Code, Pull Requests, and Off-topic.

TierneyC committed:
Add Contribution Guidelines

Added contribution guidelines: Code, PRs, and Off-Topic.

TierneyC committed:
Merge branch 'master' of github.com:iojs/website

snostorm commented:
Since most of our PRs here deal with content and internationalization efforts it might be good to build that in here as well.

We have a "code" section, which is good (some code is in this repo), but we should probably outline some content suggestions for idealized PRs.

bnb commented:
@snostorm Do you have any main points you'd like to be added to the content and internationalization section?

snostorm commented:
Putting this on the WG Agenda, we'll either merge as-is (or discuss what to add) Monday.

bnb commented:
@snostorm Sounds good. I'll be happy to make any changes that are decided on, if you guys want me to.

snostorm commented:
Content issues somewhat covered by https://github.com/iojs/evangelism/blob/master/socialMedia.styleGuide.md so maybe we can merge with some notes about that

snostorm commented:
So, in the meeting we decided that although we want to add to this, we'll just merge as-is and hope not to forget to add more.

Can we get some WG +1's still on here to make it more official since this is a `CONTRIBUTING.md` change?

Fishrock123 commented on CONTRIBUTING.md:
```diff
@@ -15,6 +15,22 @@ By making a contribution to this project, I certify that:
+* __Two contributor votes for merge:__ Two WG members of Contributor status will need to approve, or +1, your contribution in order for it to be accepted and merged into the main repo.
+
+#### Off-topic
+* __Off topic PRs:__ Please, avoid posting off-topic issues and PRs.
```
Do we actually need this bit?

snostorm replied:
Ya, if people are opening it, they probably assume it is "on topic". If some sort of issue becomes ... problematic ... then we could specifically address those issue types as part of our docs, with references on where to go instead.

Fishrock123 commented on CONTRIBUTING.md:
```diff
@@ -15,6 +15,22 @@ By making a contribution to this project, I certify that:
+#### Pull Requests
+* __PR Timeout__: If a PR has merge conflicts or one or more downvotes and the author of a PR doesn't respond to inquiries or requests posted to the PR issue after 72 hours, the PR will be closed.
+* __Controversial PRs:__ If a PR is controversial, it will be discussed by Contributors in the issue, and, if required, in the bi-weekly Website Working Group (WG) meeting.
+* __Two contributor votes for merge:__ Two WG members of Contributor status will need to approve, or +1, your contribution in order for it to be accepted and merged into the main repo.
```
I'm still less-than-sure of this 2x +1 thing

snostorm replied:
I'm not sure what "Two WG members of Contributor status" means. Anybody who has write access to the repo? We're a little slow to review/comment lately to meet this requirements regardless.

I think it is a bit of a judgement call. WG minutes, policy changes, big layout changes, code/process changes, etc. it is good for multiple. For small content tweaks/obvious stuff, seems silly.

fhemberger replied:
To keep it more general (and get it merged), why not drop the "two" and just go with this instead:

> - **Contributor votes for merge:** WG members of Contributor status will need to approve, or +1, your contribution in order for it to be accepted and merged into the main repo.

It still explains the process without being too fixated on 1/2/n approvals. We can still make this more explicit later on (if needed).

Fishrock123 commented on CONTRIBUTING.md:
```diff
@@ -15,6 +15,22 @@ By making a contribution to this project, I certify that:
+While you're working on contributions, and submitting pull requests (PRs) to add those contributions, to this project, please keep these guidelines in mind. We _expect_ submitted code, and its associated PR, to follow these standards. Please follow them so you don't have to lose your time fixing your PRs, and so we don't have to lose time explaining why.
+
+#### Code
+* __One function per PR:__ Please add only one function or feature with one PR. The project is easier to maintain when we can easily grasp everything a PR is doing, and can reference each feature by individual PR in the future.
```
Perhaps "one topic per PR"?

snostorm replied:
Also not too happy with this wording. Part of the reason I originally flagged this for the last TC. We should try and resolve any little things before the next one though. (Or at least leave some "holes" on anything we're not 100% on.)

arthurvr commented on CONTRIBUTING.md:
```diff
@@ -15,6 +15,22 @@ By making a contribution to this project, I certify that:
+
+While you're working on contributions, and submitting pull requests (PRs) to add those contributions, to this project, please keep these guidelines in mind. We _expect_ submitted code, and its associated PR, to follow these standards. Please follow them so you don't have to lose your time fixing your PRs, and so we don't have to lose time explaining why.
+
+#### Code
```
We may also mention that the code should be consistent with the code we already have. EOLs and code style for example.

fhemberger replied:
A link to the editorconfig-website should go along with this as well.

fhemberger commented on CONTRIBUTING.md:
```diff
@@ -15,6 +15,22 @@ By making a contribution to this project, I certify that:
+* __One function per PR:__ Please add only one function or feature with one PR. The project is easier to maintain when we can easily grasp everything a PR is doing, and can reference each feature by individual PR in the future.
+
+#### Pull Requests
+* __PR Timeout__: If a PR has merge conflicts or one or more downvotes and the author of a PR doesn't respond to inquiries or requests posted to the PR issue after 72 hours, the PR will be closed.
```
As it often takes some time for a PR to be reviewed/commented, I'd love to drop the "after 72 hours". Also there are people contributing on their weekends which might be too busy to reply in a time frame that close.

fhemberger commented:
@bnb Can you please look into the comments, this PR has been around for so long, it would be great if we finally could get it merged.

TierneyC committed:
Update CONTRIBUTING.md to resolve comments made.

bnb commented:
Sorry for forgetting about this - I try to be good about issues I create, but i didn't follow through with this one. If I did it properly, I addressed all comments made on this issue.

fhemberger commented:
Besides the little duplication above, it LGTM. :shipit:

Commit:
Fixed grammar error.

bnb commented:
@fhemberger Fixed that. Merging.
